### ISSUE TITLE
Enhance security of rapid-test proof feature

### ISSUE THREAD

**Issue by Ein-Tim** [corona-warn-app/cwa-wishlist#534]
## Current Implementation
Since version 2.1 it is possible to register rapid antigen tests into the app. 
Besides of the warning functionality for positive tests, which we already know from PCR-Tests, negative RAT can also be used as a proof that a rapid test was done and it returned negative. 
To use the proof feature, you have to allow that the participating test station transfers your name as well as your date of birth into the Corona-Warn-App.
If you now want to, e.g. enter a restaurant, you can proof the negative rapid antigen test with the Corona-Warn-App, which will show a screen like this: 
<img width="300" alt="image" src="https://user-images.githubusercontent.com/67682506/119697643-b481d200-be50-11eb-8593-35b672462545.png">

## Problem
Currently, the app provides no cryptographic proof that certifies the authenticity of the test result.
Everybody could build a (web-)application which looks like the Corona-Warn-App and always shows a negative rapid antigen test.
As a demonstration you can take a look at [this website](https://eyecatchup.github.io/cwa-testnachweis/) which does this, you can enter your name and your date of birth and it gets nearly impossible to identify the “fake”.

## Threat
This is a threat since people who were not tested could grant access to events, etc. that require a negative test result. 

----
Internal Tracking-ID: [EXPOSUREAPP-7543](https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-7543)

**Comment by jucktnich** [comment]
Probably use the digital pass technology

**Comment by MikeMcC399** [comment]
@jucktnich 
> Probably use the digital Green pass technology

So in that case, this wishlist item should be linked to https://github.com/corona-warn-app/cwa-wishlist/issues/424.

"A [Digital Green Certificate](https://ec.europa.eu/info/live-work-travel-eu/coronavirus-response/safe-covid-19-vaccines-europeans/covid-19-digital-green-certificates_en) is a digital proof that a person has been vaccinated against COVID-19, has recovered from COVID-19 **or has a negative test result**. The Digital Green Certificate is therefore a name used to collectively refer to 3 types of certificates: vaccination, recovery, test."

**Comment by eyecatchup** [comment]
To stress the point, here's a side-by-side comparison. No chance, for someone quick checking it, to recognize the fake.

Left: Original CWA (Android), Right: [CWA Web Clone](https://eyecatchup.github.io/cwa-testnachweis/) (installed as PWA)

<a href="https://user-images.githubusercontent.com/845131/119791503-2c91db80-bed5-11eb-867e-47c7be858ffc.png" title="Full size">
    <img src="https://user-images.githubusercontent.com/845131/119791503-2c91db80-bed5-11eb-867e-47c7be858ffc.png" alt="thumb" width="600" />
</a>
<a href="https://user-images.githubusercontent.com/845131/119791510-2dc30880-bed5-11eb-9002-15a811f1afd5.png" title="Full size">
    <img src="https://user-images.githubusercontent.com/845131/119791510-2dc30880-bed5-11eb-9002-15a811f1afd5.png" alt="thumb" width="600" />
</a>

**Comment by heinezen** [comment]
Hello everyone,

Thanks for the notice. We will discuss this topic internally and notify you about the results here on Github.

@eyecatchup In the future, please report such things directly via our Github issues or by using the corona-warn-app.opensource@sap.com address.

----
Corona-Warn-App Open Source Team

**Comment by eyecatchup** [comment]
> In the future, please report such issues directly via our Github issues

@heinezen That sounds subtly reproachful. Could you please clarify:

- What do you mean by _such issues_?
- Are you asking to _not_ create PoCs to illustrate security concerns?
- What exactly went wrong, from your point of you?

**Comment by heinezen** [comment]
@eyecatchup 

It was not meant to be reproachful, it was just meant to be a general hint. Sorry if it came across wrongly.

> Are you asking to *not* create PoCs to illustrate security concerns?

No, not at all. It is good that you want to highlight an issue and I don't doubt that you have good intentions. We also don't blame you for finding the issue or ceating the proof-of-concept. **However**, if you want us to address the problem that you highlight, we have to be notified about it in some way. And the best way to do this is to create an issue or write us an e-mail. The sooner we get to know about it, the better. Reporting also has other benefits, e.g. that the community can discuss the problem and, since you found the issue, we can ask you for advice and more details.

----
Corona-Warn-App Open Source Team

**Comment by eyecatchup** [comment]
Thanks for clarifying, @heinezen.

> **However**, if you want us to address the problem that you highlight, we have to be notified 

I agree on that. **However**, I never cared too much if you (realize and) address the problem.

**Comment by eyecatchup** [comment]
Sorry, that might sound a bit harsh. But because In spent too much wasteless time reporting things on FOSS projects and  the low severity of this, I opted too not invest the time to report this specific issue.

Still I'm happy @Ein-Tim took care of it.

**Comment by fynngodau** [comment]
The team is currently working on adding a "Nachweise" tab to version 2.4: https://github.com/corona-warn-app/cwa-app-ios/pull/2796

Here's some screenshots from the screens that give an introduction for adding tests to the tab (of course, all **not to be considered finalized**):

Adding RAT :rat: tests (scrolled to top) | Adding PCR tests (scrolled to reveal more text) | Legal (extract)
---|---|---
![Screenshot_1622354873](https://user-images.githubusercontent.com/16943720/120094018-2b9fba80-c11e-11eb-8198-6427b30cf9b8.png) | ![Screenshot_1622355170](https://user-images.githubusercontent.com/16943720/120094142-e8921700-c11e-11eb-8c37-c9cde6206959.png) | ![Screenshot_1622355067](https://user-images.githubusercontent.com/16943720/120094089-9e109a80-c11e-11eb-962a-b6230f8c115d.png)

**Comment by heinezen** [comment]
This will fixed in release 2.4 as part of the general EU Digital COVID certificate infrstructure.

----
Corona-Warn-App Open Source Team

**Comment by fynngodau** [comment]
Thanks for the confirmation @heinezen.

**Comment by eyecatchup** [comment]
@heinezen So if I understand [it](https://github.com/corona-warn-app/cwa-app-android/blob/be8d241ff80ac9cc9ccf30e76de23871dc3b5454/Corona-Warn-App/src/main/res/values/strings.xml#L1034) correct, the certificate is just an optional feature? 

```xml
<string name="submission_consent_call_test_result_checkmark_text">Wenn Ihr Testergebnis negativ ist, können Sie dies in Form eines offiziellen COVID-19-Testzertifikats bestätigen lassen. Fordern Sie hierfür in den folgenden Schritten das Testzertifikat an.</string>
```

It can be requested by the user for any rapid test result, which is displayed in the app. But the result card in the dashboard view stays untouched? Or will the result only be shown upon (certification) request by the user in the future?

Thanks for clarifying.

**Comment by heinezen** [comment]
@eyecatchup 

> So if I understand it correct, the certificate is just an optional feature?
> It can be requested by the user for any rapid test result, which is displayed in the app. But the result card in the dashboard view stays untouched? Or will the result only be shown upon (certification) request by the user in the future?

We'll publish screenshots of the new feature once it's ready. They should arrive soon as every other part of the new certificate flow is currently being finalized

Requesting a certificate is optional for both PCR and rapid-tests. I should note that the submission/request flow regarding certificates is slightly different for PCR and rapid-test submission. The information whether the user can request a rapid test certificate is encoded in the submission QR code. Therefore, I expect there to be slight differences in the UI depending on whether you requested a certificate or whether a certificate can be requested at all. The dashboard might also change based on this information.

----
Corona-Warn-App Open Source Team

**Comment by heinezen** [comment]
Hello everyone,

CWA release 2.4 does support rapid-test certificates now. They should be used as proof for a negative test result. The new process is described in our blogpost.

https://www.coronawarn.app/en/blog/2021-06-24-cwa-version-2-4/ 

> At the appropriate location (e.g. in a restaurant or at the airport), those responsible must verify the certificate's validity by scanning the QR code with the CovPass Check app and checking the ID card.

----
Corona-Warn-App Open Source Team

**Comment by Ein-Tim** [comment]
@heinezen Yes, version 2.4 improved this with the optional process of requesting a EU test certificate. However, the security flaw mentioned in the OP is still there - since the proof feature inside of CWA (independent of the DCC) is still there. So this has not been fixed or addressed in version 2.4, sorry.

**Comment by eyecatchup** [comment]
I agree with @Ein-Tim. Only certified test results should be shown.

**Comment by Ein-Tim** [comment]
Please remove the `Fix 2.4` label here.

**Comment by Ein-Tim** [comment]
One more time, please remove the `Fix 2.4` label from this issue. Thanks!

cc @dsarkar & @larswmh

**Comment by Ein-Tim** [comment]
With version 2.21 the countup/timestamp was removed, the view was aligned to PCR tests results. 

Closing.
